Turned main function in prte.c into library function provided in libprrte.so

gerargarciaa opened the pull request:
Copied prte.c and made main into a library function while also renaming the copied filed to prtelauncher.c. Also removed prte.c from src/tools/prte/ while removing any references to prte.c in Makefile.am

Commit:
Reworked PRRTE to transition the functionality of prte.c into a library function and restructure any related components.

Removed prte.c and all references to it from the Makefile.am, and successfully extracted and refactored the main function from
prte.c into the library function prte_launcher.c. Also relocated prte_launcher.c to src/runtime to align with its role as a library
function rather than a tool in src/tools/prte. The directory src/tools/prte was also deleted and the build system was updated to reflect these
structural changes in PRRTE. This work is apart of the ongoing effort of integrating PRRTE functionality back into OMPI, avoiding the need to
fork/exec prterun.

Signed-off-by: Gerardo Garcia <gagarc07@louisville.edu>

hppritcha commented:
It looks like we may need to update the pmix submodule to avoid the CI failure.  @jsquyres I think this is happening because the CI is using HEAD of openpmix.  someone (i wonder who) did a bunch of refactoring of pmix/prrte that removed PMIX_SESSION_PROVISION from the pmix header file(s).

hppritcha commented:
i'm sorry - i may not be clear enough.  We need to rebase the capstone-devel prrte on top of upstream.  or we need to a patch to capstone-devel prrte to build against head of openpmix master.

hppritcha commented on src/tools/prte/Makefile.am:
```diff
@@ -39,8 +39,10 @@ AM_CFLAGS = \
 
 bin_PROGRAMS = prte
```
this line needs to be removed for distcheck to pass

hppritcha commented on include/prte.h:
```diff
@@ -21,6 +21,7 @@
 /**
  * Main body of prte functionality
  */
-int prte(int argc, char *argv[]);
```
i would suggest leaving this prototype in the header file for now in case something is using it.

gerargarciaa replied:
Hi, just for clarification would the header file include both prte and prte_launch then for the time being? 
for example:

/**
 * Main body of prte functionality
 */
int prte(int argc, char *argv[]);
int prte_launch(int argc, char *argv[]);

jsquyres replied:
Yes.

hppritcha replied:
yes keep both prototypes in the header file, ```prte``` and  ```prte_launch```

jsquyres commented:
Note that you got build errors across the board here in CI:

```
prted/pmix/pmix_server_queries.c:837:71: warning: incompatible pointer types passing 'void **' to parameter of type 'pmix_value_t **' (aka 'struct pmix_value **') [-Wincompatible-pointer-types]
                ret = PMIx_Get(&pproc, PMIX_MEM_ALLOC_KIND, &info, 1, (void**)&value);
                                                                      ^~~~~~~~~~~~~~
/Users/runner/work/_temp/pmixinstall/include/pmix.h:208:51: note: passing argument to parameter 'val' here
                                   pmix_value_t **val);
                                                  ^
prted/pmix/pmix_server_session.c:145:50: error: use of undeclared identifier 'PMIX_SESSION_PROVISION'
        } else if (PMIX_CHECK_KEY(&req->info[n], PMIX_SESSION_PROVISION) ||
                                                 ^
1 error generated.
make[2]: *** [prted/pmix/libprrte_la-pmix_server_session.lo] Error 1
```

You might want to try building before pushing up to the PR.

hppritcha commented:
No the problem here doesn't have anything to do with the capstone work.  it has to do with the fact that certain "changes" were made to the upstream openpmix that require us to either slurp up the corresponding changes in the upstream prrte master or figure out how to change the CI to use a working version of openpmix or else patch the capstone prrte devel branch so it can work with CI using head of upstream openpmix master or something.  I believe btw, that https://github.com/open-mpi/ompi/pull/12906 is failing owing to changes introduced in upstream prrte/openpmix that included the code changes that are now breaking CI for this project. So i'm reluctant to just patch both the Open MPI prrte and the capstone prrte with those changes in upstream prrte/openpmix.

hppritcha commented:
One idea to try folks is to modify the .github/workflows yml files to build against a specific sha of openpmix.
See https://stackoverflow.com/questions/73879978/how-to-create-a-github-action-to-checkout-a-specific-commit-in-a-private-reposit

You would want to use the same openpmix sha as what we are currently using in Open MPI for main branch.
See https://github.com/open-mpi/ompi/tree/main/3rd-party.   looks like a sha to try is [08e41ed](https://github.com/openpmix/openpmix/tree/08e41ed5629b51832f5708181af6d89218c7a74e)

rhc54 commented:
> No the problem here doesn't have anything to do with the capstone work. it has to do with the fact that certain "changes" were made to the upstream openpmix that require us to either slurp up the corresponding changes in the upstream prrte master or figure out how to change the CI to use a working version of openpmix or else patch the capstone prrte devel branch so it can work with CI using head of upstream openpmix master or something.

Yes, PRRTE has always tracked with PMIx. However, I have gone through some effort to "protect" the code so that PRRTE will build with earlier versions. I'm afraid you just haven't kept up with the changes 🤷‍♂️ Current PRRTE head of master branch will build with older PMIx versions down to PMIx v4.2

hppritcha committed:
GitHub Actions: checkout specific pmix sha to use for CI

add fetch depth 0

Till we figure out what got busted in upstream pmix/prrte combo.
See what's happening with

https://github.com/open-mpi/ompi/pull/12906

Signed-off-by: Howard Pritchard <howardp@lanl.gov>

hppritcha commented:
okay looks like one more fixed needed for ```make distcheck```

```
ERROR: files left after uninstall:
./lib/libprrte.so.4.0.0
make[1]: *** [Makefile:1026: distuninstallcheck] Error 1
make[1]: Leaving directory '/home/runner/work/prrte/prrte/prrte-gitclone/_build/sub'
make: *** [Makefile:973: distcheck] Error 1
```

gerargarciaa committed:
Merge pull request #8 from uofl-capstone-open-mpi/pr/7

GitHub Actions: checkout specific pmix sha to use for CI

jsquyres reviewed (approved):
@gagarc Go ahead and merge
